### PR TITLE
Fix passing pointers as call arguments

### LINKED ISSUE #735: Errors on passing pointers as an argument to function calls
Clang reports errors on clad generated code when pointers are passed as arguments to function calls.

Reproducible example:
```cpp
#include "clad/Differentiator/Differentiator.h"

void do_nothing(const double *v, double *u) { }

double fn(double u, double v) {
    double *pv = &v;
    double *pu = &u;
    do_nothing(pv, pu);
    return v;
}

int main() {
    auto fn_grad = clad::gradient(fn);
}
```

**Errors:**
All errors are similar to:
```
 error: no viable conversion from 'double *' to 'clad::array<double>'
```

### PR DESCRIPTION
**Pull request by vaithak**
fixes #735, #636

### CODE REVIEW

**Comment by vgvassilev** [comment]
Can you check if this PR fixes #762 by any chance?

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 1]
```diff
@@ -1475,8 +1483,11 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
         // ```
         // _r0 = _d_a;
         // ```
-        Expr* dArg = nullptr;
-        if (utils::isArrayOrPointerType(argDiff.getExpr()->getType())) {
+        if (argDiff.getExpr_dx() && utils::IsLiteral(argDiff.getExpr_dx())) {
+          dArg = StoreAndRef(argDiff.getExpr_dx(), arg->getType(),
+                             direction::reverse, "_r",
+                             /*forceDeclCreation=*/true);
+        } else if (argDiff.getExpr()->getType()->isArrayType()) {
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1490]: We should refactor `argDiff.getExpr()->getType()` into a temporary to improve readibility.

**Comment by vaithak** [comment]
> Can you check if this PR fixes #762 by any chance?

That is because of a separate reason, commented on that issue.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 2]
```diff
@@ -1536,7 +1551,8 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
       // FIXME: We cannot use GlobalStoreAndRef to store a whole array so now
       // arrays are not stored.
       StmtDiff argDiffStore;
-      if (passByRef && !argDiff.getExpr()->getType()->isArrayType())
+      if (passByRef && !argDiff.getExpr()->getType()->isArrayType() &&
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1554]: Can't we use `argDiffType` here, too?

Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp]: It was inside the if condition earlier, but moved it out now 👍🏼

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Comment by parth-07** [comment]
@vaithak Can you please update pull request description / commit message with details how it addresses the issue?

Does this pull request addresses pointer arithmetic expressions such as `p + 1` used as function call arguments? If so, can you please add a test?

Also, wouldn't this pull request https://github.com/vgvassilev/clad/issues/762 fix the issue as well?

**Comment by vgvassilev** [comment]
@parth-07, I did not see your comment before merging. I guess I was too excited about this PR. Apologies!

**Comment by vaithak** [comment]
> @vaithak Can you please update pull request description / commit message with details how it addresses the issue?
> 
> Does this pull request addresses pointer arithmetic expressions such as `p + 1` used as function call arguments? If so, can you please add a test?
> 
> Also, wouldn't this pull request #762 fix the issue as well?

For #762, I have commented more about that issue itself. Regarding pointer arithmetic, I will test it out and make changes if required along with test cases in a separate PR.

**Pull request merged**